code/controllers/subsystem/garbage.dm

```diff
@@ -234,14 +237,34 @@ SUBSYSTEM_DEF(garbage)
 	queue[++queue.len] = list(gctime, refid) // not += for byond reasons
 
 //this is mainly to separate things profile wise.
-/datum/controller/subsystem/garbage/proc/HardDelete(datum/D)
+/datum/controller/subsystem/garbage/proc/HardDelete(datum/D, non_datum = FALSE, override = FALSE)
+	if(!D)
+		return
+
+	if (!override && !enable_hard_deletes)
+		return
```

Comment:
You can save yourself all this extra code and logic by instead checking this at gc_failure time,

My pr, if you'll notice, makes the distinction between ignoring gc failures, and ignoring hard deletes.

It might be better to just allow the former, and always respect things that request a hard delete.

Reply:
MSO is right, we should always respect the qdel hint, not doing so is a recipe for pain